pennylane/circuit_drawer.py

```diff
@@ -0,0 +1,916 @@
+
+    def element_representation(self, element, wire):
+        """Resolve the representation of an element in the circuit's Grid.
+
+        Args:
+            element (Union[NoneType,str,qml.operation.Operator]): The circuit element whos representation shall be resolved
+            wire (int): The element's wire which is currently resolved
+
+        Returns:
+            str: String representation of the element
+        """
+        if element is None:
+            return ""
+        if isinstance(element, str):
+            return element
+        if isinstance(element, qml.operation.Observable) and element.return_type is not None:
+            return self.output_representation(element, wire)
+
+        return self.operator_representation(element, wire)
+
+
+def _remove_duplicates(input_list):
+    """Remove duplicate entries from a list.
+
+    This operation preserves the order of the list's elements.
+
+    Args:
+        input_list (list[Hashable]): The list whose duplicate entries shall be removed
+
+    Returns:
+        list[Hashable]: The input list without duplicate entries
+    """
+    return list(OrderedDict.fromkeys(input_list))
+
+
+class CircuitDrawer:
+    """Creates a circuit diagram from the operators of a CircuitGraph in grid form.
+
+    Args:
+        raw_operation_grid (list[list[qml.operation.Operation]]): The CircuitGraph's operations
+        raw_observable_grid (list[list[qml.operation.Observable]]): The CircuitGraph's observables
+        charset (pennylane.circuit_drawer.CharSet, optional): The CharSet that shall be used for drawing. Defaults to UnicodeCharSet.
+        show_variable_names (bool, optional): Show variable names instead of variable values. Defaults to False.
+    """
+
+    def resolve_representation(self, grid, representation_grid):
+        """Resolve the string representation of the given Grid.
+
+        Args:
+            grid (pennylane.circuit_drawer.Grid): Grid that holds the circuit information
+            representation_grid (pennylane.circuit_drawer.Grid): Grid that is used to store the string representations
+        """
+        for i in range(grid.num_layers):
+            representation_layer = [""] * grid.num_wires
+
+            for wire, operator in enumerate(grid.layer(i)):
+                representation_layer[wire] = self.representation_resolver.element_representation(
+                    operator, wire
+                )
+
+            representation_grid.append_layer(representation_layer)
+
+    def resolve_decorations(self, grid, representation_grid, separate):
+        """Resolve the decorations of the given Grid.
+
+        Args:
+            grid (pennylane.circuit_drawer.Grid): Grid that holds the circuit information
+            representation_grid (pennylane.circuit_drawer.Grid): Grid that holds the string representations and into
+                which the decorations will be inserted
+            separate (bool): Insert decorations into separate layers
```

Comment:
As per my current understanding, the following sentence would further elaborate on this (if so, could be added to the docstring, or something similar in content):
```
If ``True``, a separate layer containing the decorator is created for each unique layer operator.

Reply:
I added a clarification.